### PR TITLE
rm_config: add device removal support for nmstate provider

### PR DESCRIPTION
**Pull request by abhiramnarayana**
This patch adds support for removing network devices through the nmstate implementation, extending the device lifecycle management capabilities. This includes proper error handling, type conversion, and graceful handling of non-existent interfaces.

The implementation adds _process_device_removal() method to handle device removal operations with comprehensive error logging and exception handling. A new _remove_nmstate_interface_with_type() method manages the actual interface removal using nmstate APIs, while _convert_to_nmstate_type() provides mapping between os-net-config device types and nmstate interface types.

Enhanced test coverage includes scenarios for successful device removal, error handling during removal failures, and edge cases such as removing non-existent interfaces.

Assisted-by: cursor/claude-4-sonnet

### CODE REVIEW

File: os_net_config/impl_nmstate.py [hunk 1]
```diff
@@ -2736,3 +2737,152 @@ def apply(self, cleanup=False, activate=True, config_rules_dns=True):
...
+        """Clean up SR-IOV configuration for an interface if applicable.
+
+        Uses the SR-IOV cleanup APIs from the dependent patches.
+
+        :param iface_name: Interface name to check and clean up
+        """
+        try:
+            # Check if this is an SR-IOV PF by trying to reset it
+            # This will only succeed if it's actually an SR-IOV PF
+            sriov_config.reset_sriov_pf(iface_name)
```
[the remark this answers is not in the document]
Comment (abhiramnarayana) [reply os_net_config/impl_nmstate.py:2879]: okay. will remove this.

File: os_net_config/impl_nmstate.py [hunk 2]
```diff
@@ -2736,3 +2737,152 @@ def apply(self, cleanup=False, activate=True, config_rules_dns=True):
...
+
+            # If we get here, it was an SR-IOV PF, so clean up the mappings
+            utils.remove_sriov_entries_for_pf(iface_name)
```
[the remark this answers is not in the document]
Comment (abhiramnarayana) [reply os_net_config/impl_nmstate.py:2882]: Sure. I will remove this as we are not raising any exception here. but just print error which will get printed in a different place however.

**Review by karthiksundaravel — CHANGES_REQUESTED** [review]

File: os_net_config/impl_nmstate.py [hunk 3]
```diff
@@ -2853,3 +2864,95 @@ def _get_dpdk_port_pci_address(self, dpdk_port_name):
...
+        # Process devices in specific order
+        processing_order = ['ovs_dpdk_port', 'ovs_dpdk_bond', 'ovs_bond',
+                            'vlan', 'interface', 'sriov_vf', 'ovs_bridge',
+                            'ovs_user_bridge', 'linux_bond', 'sriov_pf']
+
+        for device_type in processing_order:
+            devices_of_type = [device for device in remove_device_list
+                               if device.remove_type == device_type]
+
+            if devices_of_type:
```
Comment (karthiksundaravel) [inline os_net_config/impl_nmstate.py:2908]: This 'if' statement is taken care by the 'for' loop, so it shall be removed

File: os_net_config/impl_nmstate.py [hunk 4]
```diff
@@ -2853,3 +2864,95 @@ def _get_dpdk_port_pci_address(self, dpdk_port_name):
...
+            logger.warning("%s: device is managed by nmstate provider, but "
+                           "cannot "
+                           "be removed due to a probable dependency on other "
+                           "devices", device.remove_name)
+            device.is_removed = True
+            return
+
+        # Step 1: Get PCI addresses for DPDK devices (before cleaning)
+        if device.remove_type in ['ovs_dpdk_port', 'ovs_dpdk_bond'] and \
+            device.provider_data is not None:
```
Comment (karthiksundaravel) [inline os_net_config/impl_nmstate.py:2933]: The validation 'device.provider_data is not None' is already taken care in line 2923, so its redundant here

File: os_net_config/impl_nmstate.py [hunk 5]
```diff
@@ -2853,3 +2864,95 @@ def _get_dpdk_port_pci_address(self, dpdk_port_name):
...
+            pci_addresses = device.provider_data.pci_address
+        else:
+            pci_addresses = []
+
+        # Step 2: Remove interface using nmstate with type conversion
+        if device.provider_data is not None and \
```
Comment (karthiksundaravel) [inline os_net_config/impl_nmstate.py:2939]: same as above.

**Review by karthiksundaravel — COMMENTED** [review]
Some minor comments

File: os_net_config/impl_nmstate.py (hunk 4 above)
Comment (abhiramnarayana) [reply os_net_config/impl_nmstate.py:2933]: Sure. Will update accordingly.

File: os_net_config/impl_nmstate.py [hunk 6]
```diff
@@ -2853,3 +2864,89 @@ def _get_dpdk_port_pci_address(self, dpdk_port_name):
...
+        if device.provider_data.dev_type is not None:
+            self._clean_iface(
+                device.remove_name, device.provider_data.dev_type
+            )
+
+        # Step 3: Clean up DPDK interfaces or SR-IOV entries
+        if device.remove_type in ['ovs_dpdk_port', 'ovs_dpdk_bond']:
+            # For DPDK: use the PCI addresses we collected earlier
+            for pci_address in pci_addresses:
+                utils.remove_dpdk_interface(pci_address)
```
[the remark this answers is not in the document]
Comment (abhiramnarayana) [reply os_net_config/impl_nmstate.py:2943]: Sure. Will add this.

**Review by karthiksundaravel — DISMISSED** [review]
LGTM

**Review by karthiksundaravel — APPROVED** [review]
LGTM
